File: download_flexgate.py
# -*- coding: utf-8 -*-
import sys
sys.stdout.reconfigure(encoding="utf-8")
sys.stderr.reconfigure(encoding="utf-8")
"""
download_flexgate.py
flexgate 로그인 → 배송준비 주문 목록 → 엑셀 다운로드

사용: python download_flexgate.py <저장경로>
예시: python download_flexgate.py D:\경락가데이터서버\eomgung-market\order_output
"""
import sys, re, json, time, os
from datetime import datetime
import urllib.request, urllib.parse, urllib.error
import http.cookiejar

BASE   = 'https://dongnaegotgan.flexgate.co.kr'
INTRO  = 'https://intro.flexgate.co.kr'
UID    = 'dongnaegotgan'
UPW    = '곳간12!@'

def log(msg):
    print(f"[{datetime.now().strftime('%H:%M:%S')}] {msg}", flush=True)

def make_opener():
    cj = http.cookiejar.CookieJar()
    opener = urllib.request.build_opener(urllib.request.HTTPCookieProcessor(cj))
    opener.addheaders = [
        ('User-Agent','Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0.0.0 Safari/537.36'),
        ('Accept','text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8'),
        ('Accept-Language','ko-KR,ko;q=0.9,en;q=0.8'),
    ]
    return opener, cj

def login(opener):
    log('로그인 페이지 로드...')
    res = opener.open(f'{INTRO}/Mypage/Login')
    html = res.read().decode('utf-8')

    # encKey 추출
    m = re.search(r'name=["\']encKey["\'][^>]*value=["\']([^"\']*)["\']', html)
    enc_key = m.group(1) if m else ''
    log(f'encKey: {enc_key[:20]}...' if len(enc_key) > 20 else f'encKey: {repr(enc_key)}')

    # 로그인 POST
    login_data = urllib.parse.urlencode({
        'userId': UID,
        'password': UPW,
        'encKey': enc_key,
        'returnUrl': '/NewOrder/deal01?order_status=30&formtype=A&pagesize=1000'
    }).encode('utf-8')

    req = urllib.request.Request(
        f'{INTRO}/Mypage/Login',
        data=login_data,
        headers={'Content-Type': 'application/x-www-form-urlencoded',
                 'Referer': f'{INTRO}/Mypage/Login'},
        method='POST'
    )
    res = opener.open(req)
    final_url = res.geturl()
    log(f'로그인 후 URL: {final_url[:80]}')
    return final_url

def get_order_numbers(opener):
    log('배송준비 주문 목록 로드...')
    url = f'{BASE}/NewOrder/deal01?order_status=30&formtype=A&pagesize=1000'
    res = opener.open(url)
    html = res.read().decode('utf-8')

    # 주문번호 추출: input[name="chk"] value="..." data-mogidx="..."
    # 실제로는 mo_order_num = mogidx 값들
    mogidx_list = re.findall(r'data-mogidx=["\'](\d+)["\']', html)
    
    if not mogidx_list:
        # 다른 패턴 시도
        mogidx_list = re.findall(r'name=["\']chk["\'][^>]*value=["\']([^"\']+)["\']', html)
    
    log(f'주문 건수: {len(mogidx_list)}건')
    return mogidx_list

def create_excel(opener, mogidx_list):
    log('엑셀 파일 생성 요청...')
    post_data = urllib.parse.urlencode({
        'mo_order_num': ','.join(mogidx_list),
        'types': '3',
        'customno': '94'
    }).encode('utf-8')

    req = urllib.request.Request(
        f'{BASE}/NewOrder/CreateExcelIfile',
        data=post_data,
        headers={
            'Content-Type': 'application/x-www-form-urlencoded',
            'Referer': f'{BASE}/NewOrder/deal01?order_status=30&formtype=A&pagesize=1000',
            'X-Requested-With': 'XMLHttpRequest'
        },
        method='POST'
    )
    res = opener.open(req)
    resp_text = res.read().decode('utf-8')
    log(f'서버 응답: {resp_text[:200]}')

    # 파일명 추출
    m = re.search(r'order_\d+\.xlsx', resp_text)
    if m:
        return m.group(0)
    
    try:
        data = json.loads(resp_text)
        return data.get('fileName') or data.get('filename') or data.get('data')
    except:
        return resp_text.strip().strip('"\'')

def download_excel(opener, file_name, out_dir):
    log(f'엑셀 다운로드: {file_name}')
    time.sleep(3)  # 서버 파일 생성 대기

    url = f'{BASE}/NewOrder/ExcelDownload?fileName={urllib.parse.quote(file_name)}'
    req = urllib.request.Request(url, headers={
        'Referer': f'{BASE}/NewOrder/deal01?order_status=30&formtype=A&pagesize=1000'
    })
    res = opener.open(req)
    
    content = res.read()
    content_type = res.headers.get('Content-Type', '')
    log(f'응답: {len(content)}bytes, {content_type}')

    if len(content) < 500 or 'html' in content_type.lower():
        # 재시도
        log('재시도...')
        time.sleep(3)
        res2 = opener.open(req)
        content = res2.read()
        log(f'재시도 응답: {len(content)}bytes')

    os.makedirs(out_dir, exist_ok=True)
    out_path = os.path.join(out_dir, file_name)
    with open(out_path, 'wb') as f:
        f.write(content)
    
    log(f'저장 완료: {out_path}')
    return out_path

def main():
    out_dir = sys.argv[1] if len(sys.argv) > 1 else '.'
    
    opener, cj = make_opener()
    
    # 1. 로그인
    final_url = login(opener)
    if 'Login' in final_url or 'login' in final_url:
        log('❌ 로그인 실패 - URL이 여전히 로그인 페이지')
        sys.exit(1)
    log('✅ 로그인 성공')

    # 2. 주문번호 목록
    mogidx_list = get_order_numbers(opener)
    if not mogidx_list:
        log('❌ 주문 없음')
        sys.exit(1)

    # 3. 엑셀 생성
    file_name = create_excel(opener, mogidx_list)
    if not file_name:
        log('❌ 파일명을 받지 못했습니다')
        sys.exit(1)
    log(f'파일명: {file_name}')

    # 4. 다운로드
    out_path = download_excel(opener, file_name, out_dir)
    
    # 성공 출력 (Node.js가 읽음)
    print(f'DOWNLOAD_OK:{out_path}')

if __name__ == '__main__':
    main()

File: download_flexgate.js
/**
 * download_flexgate.js
 * 
 * 로그인 창이 열리면 → 로그인 버튼만 클릭하세요 (자동완성으로 채워짐)
 * 로그인 후 자동으로 배송준비 엑셀 다운로드
 */
require('dotenv').config({ path: require('path').join(__dirname, '.env') });
const puppeteer = require('puppeteer');
const readline  = require('readline');
const fs   = require('fs');
const path = require('path');

const BASE    = 'https://dongnaegotgan.flexgate.co.kr';
const INTRO   = 'https://intro.flexgate.co.kr';
const OUT_DIR = process.argv[2] || '.';

function log(m) { process.stdout.write(`[${new Date().toLocaleTimeString('ko-KR')}] ${m}\n`); }
function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }

async function run() {
  fs.mkdirSync(OUT_DIR, { recursive: true });

  const browser = await puppeteer.launch({
    headless: false,
    args: [
      '--no-sandbox', '--disable-setuid-sandbox',
      '--disable-blink-features=AutomationControlled',
    ],
    defaultViewport: null,
    ignoreDefaultArgs: ['--enable-automation'],
  });

  const page = (await browser.pages())[0] || await browser.newPage();
  page.on('dialog', d => d.accept().catch(() => {}));

  const cdp = await page.createCDPSession();
  await cdp.send('Page.setDownloadBehavior', { behavior: 'allow', downloadPath: path.resolve(OUT_DIR) });

  try {
    // ── 로그인 ──────────────────────────────────────────────────
    log('로그인 페이지 열기...');
    await page.goto(`${INTRO}/Mypage/Login`, { waitUntil: 'domcontentloaded', timeout: 20000 });
    await sleep(1500);

    log('');
    log('================================================================');
    log('  브라우저 창에서 로그인 버튼을 클릭해주세요!');
    log('  (아이디/비번은 자동완성으로 이미 채워져 있어요)');
    log('================================================================');
    log('');

    // 로그인 완료 대기 (URL이 dongnaegotgan.flexgate.co.kr로 바뀔 때까지)
    try {
      await page.waitForFunction(
        () => location.hostname === 'dongnaegotgan.flexgate.co.kr',
        { timeout: 300000, polling: 500 } // 5분 대기
      );
    } catch(e) {
      log('❌ 5분 내 로그인 안 됨');
      await browser.close();
      process.exit(1);
    }
    log('✅ 로그인 완료');

    // ── 배송준비 페이지 이동 ────────────────────────────────────
    log('배송준비 페이지 이동...');
    await page.goto(
      `${BASE}/NewOrder/deal01?order_status=30&formtype=A&pagesize=1000`,
      { waitUntil: 'networkidle2', timeout: 30000 }
    );
    await sleep(2000);

    // 주문 로드 대기
    await page.evaluate(() => {
      if (typeof loadedFunc === 'function') loadedFunc();
    }).catch(() => {});
    await page.waitForFunction(
      () => document.querySelectorAll('input[name="chk"]').length > 0,
      { timeout: 20000, polling: 500 }
    ).catch(() => {});
    await sleep(1000);

    const cnt = await page.evaluate(() => document.querySelectorAll('input[name="chk"]').length);
    log(`주문 ${cnt}건 로드됨`);

    if (cnt === 0) {
      log('❌ 주문 없음');
      await browser.close();
      process.exit(1);
    }

    // ── 전체선택 + 명지일일배송 + 다운로드 ─────────────────────
    log('전체선택 + 명지일일배송 선택...');
    await page.evaluate(() => {
      document.getElementById('chkCheckDataAll').click();
      document.getElementById('customexcelFrm').value = '94';
    });
    await sleep(500);

    // CreateExcelIfile 응답 캡처
    const createPromise = page.waitForResponse(
      r => r.url().includes('CreateExcelIfile'), { timeout: 30000 }
    );
    await page.evaluate(() => orderExcelDownload(3));

    let fileName = null;
    try {
      const res  = await createPromise;
      const text = await res.text();
      log(`서버 응답: ${text.slice(0, 100)}`);
      const m = text.match(/order_\d+\.xlsx/);
      if (m) fileName = m[0];
      else { try { fileName = JSON.parse(text).fileName || ''; } catch(e) {} }
    } catch(e) { log(`응답 캡처: ${e.message}`); }

    if (!fileName) { log('❌ 파일명 없음'); await browser.close(); process.exit(1); }
    log(`파일명: ${fileName}`);

    // ── 다운로드 ─────────────────────────────────────────────────
    log('다운로드 중...');
    await sleep(3000);
    const dlUrl = `${BASE}/NewOrder/ExcelDownload?fileName=${encodeURIComponent(fileName)}`;
    await page.goto(dlUrl, { waitUntil: 'domcontentloaded', timeout: 20000 }).catch(() => {});
    await sleep(5000);

    // 파일 확인
    const files = fs.readdirSync(OUT_DIR)
      .filter(f => f.endsWith('.xlsx') && !f.includes('소분') && !f.includes('배송주소'))
      .map(f => ({ name: f, t: fs.statSync(path.join(OUT_DIR, f)).mtimeMs }))
      .sort((a, b) => b.t - a.t);

    if (files.length > 0 && Date.now() - files[0].t < 60000) {
      const out = path.join(OUT_DIR, files[0].name);
      log(`✅ 다운로드 완료: ${files[0].name}`);
      await browser.close();
      console.log(`DOWNLOAD_OK:${out}`);
    } else {
      log('❌ 파일 없음');
      await browser.close();
      process.exit(1);
    }

  } catch(e) {
    log(`❌ 오류: ${e.message}`);
    await browser.close().catch(() => {});
    process.exit(1);
  }
}

run();